javalib/src/main/scala/java/util/Arrays.scala

```diff
@@ -80,121 +80,68 @@ object Arrays {
   private def sortRangeImpl[@specialized T: ClassTag](
       a: Array[T], fromIndex: Int, toIndex: Int)(implicit ord: Ordering[T]): Unit = {
     checkIndicesForCopyOfRange(a.length, fromIndex, toIndex)
-    quickSort[T](a, fromIndex, toIndex)
+    stableMergeSort[T](a, fromIndex, toIndex)
   }
 
   @inline
   private def sortRangeAnyRefImpl(a: Array[AnyRef], fromIndex: Int, toIndex: Int)(
       implicit ord: Ordering[AnyRef]): Unit = {
     checkIndicesForCopyOfRange(a.length, fromIndex, toIndex)
-    quickSortAnyRef(a, fromIndex, toIndex)
+    stableMergeSortAnyRef(a, fromIndex, toIndex)
   }
 
   @inline
   private def sortImpl[@specialized T: ClassTag: Ordering](a: Array[T]): Unit =
-    quickSort[T](a, 0, a.length)
+    stableMergeSort[T](a, 0, a.length)
 
   @inline
   private def sortAnyRefImpl(a: Array[AnyRef])(implicit ord: Ordering[AnyRef]): Unit =
-    quickSortAnyRef(a, 0, a.length)
+    stableMergeSortAnyRef(a, 0, a.length)
 
-  // Implementation of sorting based on Scala 2.11.7 scala.util.Sorting
-  private final val qSortThreshold = 16
+  private final val inPlaceSortThreshold = 16
 
-  /** Sort array `a` with quicksort, using the Ordering on its elements.
-    * This algorithm sorts in place, so no additional memory is used aside from
-    * what might be required to box individual elements during comparison.
-    */
-  @noinline
-  private def quickSort[@specialized K](a: Array[K], i0: Int, iN: Int)(implicit ord: Ordering[K]): Unit = {
-    if (iN - i0 < qSortThreshold) {
+  /** Sort array `a` with merge sort and insertion sort,
+   *  using the Ordering on its elements.
+   */
+  @inline
+  private def stableMergeSort[@specialized K: ClassTag](a: Array[K], i0: Int, iN: Int)(
+      implicit ord: Ordering[K]): Unit = {
+    if (iN - i0 > inPlaceSortThreshold)
+      stableSplitMerge(a, new Array[K](a.length), i0, iN)
+    else
       insertionSort(a, i0, iN)
+  }
+
+  @noinline
+  private def stableSplitMerge[@specialized K](a: Array[K], b: Array[K],
+      i0: Int, iN: Int)(implicit ord: Ordering[K]): Unit = {
+    val length = iN - i0
+    if (length > inPlaceSortThreshold) {
+      val iM = i0 + (length / 2)
```

Comment:
Use `(i0 + iN) >>> 1 // Unsigned div by 2` which does not have the overflow problem.